server/src/main/java/org/apache/druid/segment/indexing/DataSchema.java

```diff
@@ -457,4 +413,108 @@ public String toString()
            ", inputRowParser=" + inputRowParser +
            '}';
   }
+
+  public static class Builder
+  {
+    private String dataSource;
+    private AggregatorFactory[] aggregators;
+    private GranularitySpec granularitySpec;
+    private TransformSpec transformSpec;
+    private Map<String, Object> parserMap;
+    private ObjectMapper objectMapper;
+
+    // The below fields can be initialized lazily from parser for backward compatibility.
+    private TimestampSpec timestampSpec;
+    private DimensionsSpec dimensionsSpec;
+
+    public Builder()
+    {
+
+    }
+
+    public Builder(DataSchema schema)
+    {
+      this.dataSource = schema.dataSource;
+      this.aggregators = schema.aggregators;
+      this.granularitySpec = schema.granularitySpec;
+      this.transformSpec = schema.transformSpec;
+      this.parserMap = schema.parserMap;
+      this.objectMapper = schema.objectMapper;
+      this.timestampSpec = schema.timestampSpec;
+      this.dimensionsSpec = schema.dimensionsSpec;
+    }
+
+    public Builder withDataSource(String dataSource)
+    {
+      this.dataSource = dataSource;
+      return this;
+    }
+
+    public Builder withTimestamp(TimestampSpec timestampSpec)
+    {
+      this.timestampSpec = timestampSpec;
+      return this;
+    }
+
+    public Builder withDimensions(DimensionsSpec dimensionsSpec)
+    {
+      this.dimensionsSpec = dimensionsSpec;
+      return this;
+    }
+
+    public Builder withDimensions(List<DimensionSchema> dimensions)
+    {
+      this.dimensionsSpec = DimensionsSpec.builder().setDimensions(dimensions).build();
+      return this;
+    }
+
+    public Builder withDimensions(DimensionSchema... dimensions)
+    {
+      return withDimensions(Arrays.asList(dimensions));
+    }
+
+    public Builder withAggregators(AggregatorFactory... aggregators)
+    {
+      this.aggregators = aggregators;
+      return this;
+    }
+
+    public Builder withGranularity(GranularitySpec granularitySpec)
+    {
+      this.granularitySpec = granularitySpec;
+      return this;
+    }
+
+    public Builder withTransform(TransformSpec transformSpec)
+    {
+      this.transformSpec = transformSpec;
+      return this;
+    }
+
+    public Builder withObjectMapper(ObjectMapper objectMapper)
```

Comment:
Include `@Deprecated`, since this is only used with the `parserMap`, which is itself deprecated?